edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ServiceIdentityTree.cs

```diff
@@ -0,0 +1,240 @@
+                        // We've reached the root device, we're done
+                        return true;
+                    }
+
+                    // Move on to the next parent
+                    treeNode = parent;
+                }
+
+                // If we walked the parent chain up to a dead-end, i.e. topmost node does not
+                // match tree root device ID, then it means this subtree is dangling.
+                // This is a possible scenario, we just treat it as if there's no auth chain.
+            }
+
+            return false;
+        }
+
+        void InsertModuleIdentity(ServiceIdentity module)
+        {
+            var newNode = new ServiceIdentityTreeNode(module);
+            this.nodes.Add(module.Id, newNode);
+
+            if (this.nodes.TryGetValue(module.DeviceId, out ServiceIdentityTreeNode parentDeviceNode))
+            {
+                // Hook up the module to the parent device
+                newNode.Parent = Option.Some(parentDeviceNode);
+                parentDeviceNode.Children.Add(newNode);
+            }
+            else
+            {
+                // It's okay if we can't find the parent device, it's possible that the identity
+                // for the parent device hasn't came in yet. In this case we'll just leave the
+                // module as a dangling node, and the parent should come in at a later point.
+            }
+        }
+
+        void InsertDeviceIdentity(ServiceIdentity device)
+        {
+            var newNode = new ServiceIdentityTreeNode(device);
+            this.nodes.Add(device.Id, newNode);
+
+            // Check if there's an existing parent for this device
+            device.ParentScopes.ForEach(parentScopeId =>
+            {
+                // Look for the parent based on scope ID, it's okay if we can't find a parent
+                // node. It's possible that the identity for the parent hasn't came in yet.
+                // In this case we'll just leave the new identity as a dangling node, and
+                // the parent device identity should come in at a later point.
+                Option<ServiceIdentityTreeNode> parentNode = this.FindDeviceByScopeId(parentScopeId);
+
+                // Hook up the new node into the tree
+                newNode.Parent = parentNode;
+                parentNode.ForEach(node => node.Children.Add(newNode));
+            });
+
+            // Check if there are any dangling children that can now be hooked up,
+            // this include placing the new node as the new root.
+            List<ServiceIdentityTreeNode> danglingChildren =
+                this.nodes
+                .Select(kvp => kvp.Value)
+                .Where(s => s.Identity.ParentScopes.HasValue && s.Identity.ParentScopes.Equals(device.DeviceScope))
+                .ToList();
+
+            foreach (ServiceIdentityTreeNode child in danglingChildren)
+            {
+                child.Parent = Option.Some(newNode);
```

Comment:
Np: Code is when parent comes in late is reapeate: see lines [150, 151], [124,125], [164,165].
Maybe encapsulate in a function to make sure that if one area is modified they are modified the same way.
ie: The operation that would have been done if the parent was there need to be done when the parent might finally coming. Someone changing the code in the year or 2 might not be aware that change need to be made similarly in those 3 areas.